plasmapy_nei/nei/nei.py

```diff
@@ -1,7 +1,1363 @@
+            raise NEIError(f"Unable to find new dt at t = {t}") from exc
+        else:
+            if np.isnan(new_dt.value):
+                raise NEIError(f"new_dt = {new_dt}")
+
+        # Enforce that the time step is in the interval [dt_min, dt_max].
+
+        if new_dt < self.dt_min:
+            new_dt = self.dt_min
+        elif new_dt > self.dt_max:
+            new_dt = self.dt_max
+
+        # Store the time step as a private attribute so that it can be
+        # used in the time advance.
+
+        self._dt = new_dt.to(u.s)
+
+    def set_timestep(self, dt: u.Quantity = None):
+        """
+        Set the time step for the next non-equilibrium ionization time
+        advance.
+
+        Parameters
+        ----------
+        dt: astropy.units.Quantity, optional
+            The time step to be used for the next time advance.
+
+        Notes
+        -----
+        If ``dt`` is not `None`, then the time step will be set to ``dt``.
+
+        If ``dt`` is not set and the ``adapt_dt`` attribute of an
+        `~plasmapy_nei.nei.NEI` instance is `True`, then this method will
+        calculate the time step corresponding to how long it will be
+        until the temperature rises or drops into the next temperature
+        bin.  If this time step is between ``dtmin`` and ``dtmax``, then
+
+        If ``dt`` is not set and the ``adapt_dt`` attribute is `False`,
+        then this method will set the time step as what was inputted to
+        the `~plasmapy_nei.nei.NEI` class upon instantiation in the
+        ``dt`` argument or through the `~plasmapy_nei.nei.NEI` class's
+        ``dt_input`` attribute.
+
+        Raises
+        ------
+        ~plasmapy_nei.nei.NEIError
+            If the time step cannot be set, for example if the ``dt``
+            argument is invalid or the time step cannot be adapted.
+        """
+
+        if dt is not None:
+            # Allow the time step to set as an argument to this method.
+            try:
+                dt = dt.to(u.s)
+            except Exception as exc:
+                raise NEIError(f"{dt} is not a valid time step.") from exc
+            finally:
+                self._dt = dt
+        elif self.adapt_dt:
+            try:
+                self._set_adaptive_timestep()
+            except Exception as exc:
+                raise NEIError("Unable to adapt the time step.") from exc
+        elif self.dt_input is not None:
+            self._dt = self.dt_input
+        else:
+            raise NEIError("Unable to set the time step.")
+
+        self._old_time = self._new_time
+        self._new_time = self._old_time + self._dt
+
+        if self._new_time > self.time_max:
+            self._new_time = self.time_max
+            self._dt = self._new_time - self._old_time
+
+    def time_advance(self):
+        """Advance the simulation by one time step."""
+        # TODO: Expand docstring and include equations!
+
+        # TODO: Fully implement units into this.
```

Comment:
If you want to take that performance hit, sure; but I'd try to do repetitive calculations without them for now, and keep units to interfaces. I think advancements in numpy dtypes may help, but I'm not entirely sure.

Reply:
Yeah...performance is going to be important here.

Reply:
- [ ] Create issue to optimize the time advance that includes this comment